backend/app/routes/session_utils/session_lifecycle.py

```diff
@@ -0,0 +1,161 @@
+from datetime import datetime, timedelta
+from typing import Union
+
+from beanie.odm.operators.find.comparison import In
+from beanie.odm.operators.update.general import Set
+
+from models.config import ExperimentSettings
+from models.session import Session, SessionError
+from models.subject import Subject
+from study_setup.generate_sessions import create_trials
+from utils.utils import estimate_average_player_score
+
+
+async def get_session(prolific_id) -> Union[Session, SessionError]:
+    """ Get session for the subject """
+    # check if collection Subject exists
+    if await Subject.find().count() > 0:
+        subjects_with_id = await Subject.find(
+            Subject.prolific_id == prolific_id).to_list()
+    else:
+        subjects_with_id = []
+
+    if len(subjects_with_id) == 0:
+        # subject does not exist
+        # creat a new subject
+        subject = Subject(prolific_id=prolific_id)
+        # save subject to database
+        await subject.save()
+        # session initialization for the subject
+        # session will not be assigned to the subject if there is no available
+        await initialize_session(subject)
+    elif len(subjects_with_id) > 1:
+        # if more than one subject with the same prolific id return error
+        return SessionError(
+            message='Multiple subjects with the same prolific id')
+    else:
+        subject = subjects_with_id[0]
+
+    # get session for the subject
+    session = await Session.find_one(Session.subject_id == subject.id)
+    if session is None:
+        # this happens when all available sessions are taken
+        return SessionError(message='No available session for the subject')
+
+    # this will happen only for a new subject
+    if subject.session_id is None:
+        # update Subject.session_id field if it is empty
+        await subject.update(Set({Subject.session_id: session.id}))
+
+    return session
+
+
+async def initialize_session(subject: Subject):
+    # Check and remove expired sessions
+    await replace_stale_session(ExperimentSettings())
+
+    # assign subject to any available session
+    await Session.find_one(Session.available == True).update(
```

Comment:
I'm not sure I understand why this is necessary since the parent-child connections are completely random. Could you please explain that in more detail? @LBrinkmann

Reply:
I see, thank you for the explanation! I'll fix it later.